### PR TITLE
Add logic for partial order fills.

### PR DESCRIPTION
**Pull request by kahuang**
Adds logic for partial order fills on both the buy (and sell) side

We should also probably add tests for the escrow side on the buy side

### CODE REVIEW

**Comment by kahuang** [comment]
Fixes #3 

- Refactor escrowID to be incrementing like orderID, since there can be a one-to-many mapping of order to escrow
- Allow partial order fulfillment, and calculate escrow amount proportionally (stake and amountSats)

File: packages/portal-contracts/src/Portal.sol [hunk 1]
```diff
@@ -81,6 +84,7 @@ contract Portal is Owned {
         uint256 escrowID,
         uint256 orderID,
         int128 amountSats,
+        int128 amountSatsFilled,
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:84]: Wouldn't amountSatsFilled always be zero when an order is first placed?

File: packages/portal-contracts/src/Portal.sol [hunk 2]
```diff
@@ -263,43 +271,47 @@ contract Portal is Owned {
...
         require(o.amountSats < 0, "Order already filled");
-        require(-o.amountSats == int128(amountSats), "Amount incorrect");
+        require(-o.amountSats >= int128(amountSats), "Amount incorrect");
 
         // Verify correct stake amount.
         uint256 totalTok = uint256(amountSats) * uint256(o.priceTokPerSat);
         uint256 expectedStakeTok = (totalTok * stakePercent) / 100;
 
-        // Receive stake
+        // Receive stake. Validates that msg.value == expectedStateTok (for ether based payments)
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:282]: 👍

File: packages/portal-contracts/src/Portal.sol [hunk 3]
```diff
@@ -263,43 +271,47 @@ contract Portal is Owned {
         payable
         returns (uint256 escrowID)
     {
-        // Orders can only be filled in their entirety, for now.
-        // This means escrows are 1:1 with orders.
-        // TODO: allow partial fills?
-        escrowID = orderID * 1e9;
+        escrowID = nextEscrowID++;
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:272]: 🚀

File: packages/portal-contracts/src/Portal.sol [hunk 4]
```diff
@@ -263,43 +271,47 @@ contract Portal is Owned {
...
+        // Receive stake. Validates that msg.value == expectedStateTok (for ether based payments)
         _transferFromSender(expectedStakeTok);
 
         // Put the COMBINED eth (buyer's stake + the order amount) into escrow.
         Escrow storage e = escrows[escrowID];
         e.destScriptHash = o.scriptHash;
         e.amountSatsDue = amountSats;
         e.deadline = uint128(block.timestamp + 24 hours);
-        e.escrowTok = totalTok + msg.value;
+        e.escrowTok = totalTok + expectedStakeTok;
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:290]: Good fix

File: packages/portal-contracts/src/Portal.sol [hunk 5]
```diff
@@ -263,43 +271,47 @@ contract Portal is Owned {
...
         e.successRecipient = msg.sender;
         e.timeoutRecipient = o.maker;
 
-        // Order matched and filled.
+        // Order matched.
         emit OrderMatched(
             escrowID,
             orderID,
             o.amountSats,
+            int128(amountSats),
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:299]: 👍

File: packages/portal-contracts/src/Portal.sol [hunk 6]
```diff
@@ -263,43 +271,47 @@ contract Portal is Owned {
...
             o.priceTokPerSat,
             expectedStakeTok,
             o.maker,
             msg.sender
         );
 
-        delete orderbook[orderID];
+        // Update the amount of liquidity in this order
+        o.amountSats += int128(amountSats);
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:307]: 👍

File: packages/portal-contracts/src/Portal.sol (hunk 1 above)
Comment (kahuang) [reply packages/portal-contracts/src/Portal.sol:84]: Hmm, this is the OrderMatched event, is that what you're referring to?

File: packages/portal-contracts/src/Portal.sol [hunk 7]
```diff
@@ -308,13 +320,21 @@ contract Portal is Owned {
...
-        require(o.amountSats == int128(amountSats), "Amount incorrect");
+        require(o.amountSats >= int128(amountSats), "Amount incorrect");
+
+        uint256 totalValue = amountSats * o.priceTokPerSat;
+        int256 orderSats = int256(o.amountSats).fromInt();
+        int256 fillSats = int256(uint256(amountSats)).fromInt();
+        int256 stake = int256(o.stakedTok).fromInt();
+
+        int256 ratio = fillSats.div(orderSats);
+        uint256 portionOfStake = uint256(stake.mul(ratio)) / 1e18;
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol]: we may be able to compute this without that fixed-point library, something like

```solidity
uint256 portionOfStake = o.stakedTok * uint256(amountSats) / o.amountSats;
```

To guarantee this doesn't overflow, we could enforce `o.stakedTok < 2^128` during order creation.

File: packages/portal-contracts/src/Portal.sol [hunk 8]
```diff
@@ -324,22 +344,30 @@ contract Portal is Owned {
         e.destScriptHash = destScriptHash;
         e.amountSatsDue = amountSats;
         e.deadline = uint128(block.timestamp + 24 hours);
-        e.escrowTok = o.stakedTok + msg.value;
+        e.escrowTok = portionOfStake + totalValue;
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:340]: 👍

File: packages/portal-contracts/src/Portal.sol [hunk 9]
```diff
@@ -324,22 +344,30 @@ contract Portal is Owned {
...
             o.priceTokPerSat,
             0,
             o.maker,
             msg.sender
         );
 
-        delete orderbook[orderID];
+        o.amountSats -= int128(amountSats);
+        o.stakedTok -= portionOfStake;
+        emit OrderCancelled(o.stakedTok);
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol]: We're not cancelling the order, so we don't need an `OrderCancelled` event. `OrderMatched` above looks good.

File: packages/portal-contracts/src/Portal.sol [hunk 10]
```diff
@@ -324,22 +344,30 @@ contract Portal is Owned {
...
         e.successRecipient = o.maker;
         e.timeoutRecipient = msg.sender;
 
-        // Order matched and filled.
+        // Order matched.
         emit OrderMatched(
             escrowID,
             orderID,
             o.amountSats,
+            int128(amountSats),
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:349]: 👍

File: packages/portal-contracts/src/test/Portal.t.sol [hunk 11]
```diff
@@ -147,9 +148,34 @@ contract PortalTest is Test {
...
+        uint256 escrowID = p.initiateBuy{value: 0.1 ether}(1, 1e7);
+        assertEq(escrowID, 1);
+
+        // Fill out the rest of the order
+        escrowID = p.initiateBuy{value: 0.9 ether}(1, 1e8-1e7);
+        assertEq(escrowID, 2);
+
+        // Order should be filled
+        vm.expectRevert(bytes("Order already filled"));
+        p.initiateBuy{value: 0.1 ether}(1, 1e7);
```
Comment (dcposch) [inline packages/portal-contracts/src/test/Portal.t.sol:178]: 👍 👍

File: packages/portal-contracts/src/test/Portal.t.sol [hunk 12]
```diff
@@ -198,44 +251,114 @@ contract PortalTest is Test {
...
+        p.proveSettlement(1, 123, proof, 12);
+        
+        // Finally, try again. Escrow should be gone.
+        vm.expectRevert(bytes("Escrow not found"));
+        p.proveSettlement(1, 123, proof, 12);
+        
+        // Fill the final escrow
+        vm.expectEmit(true, true, true, true);
+        emit EscrowSettled(2, 1e8-1e7, address(this), 18.9 ether);
+        p.proveSettlement(2, 123, proof, 12);
```
Comment (dcposch) [inline packages/portal-contracts/src/test/Portal.t.sol:361]: 👍

**Review by dcposch — COMMENTED** [review]
Looks excellent to me. Thanks!

I left a few comments  🚀

File: packages/portal-contracts/src/Portal.sol (hunk 9 above)
Comment (kahuang) [reply packages/portal-contracts/src/Portal.sol]: Good catch! This was a debug statement I put in before I found console.log :D

**Comment by kahuang** [comment]
@dcposch addressed your comments, PTAL!

File: packages/portal-contracts/src/Portal.sol [hunk 13]
```diff
@@ -324,22 +337,29 @@ contract Portal is Owned {
...
             msg.sender
         );
 
-        delete orderbook[orderID];
+        o.amountSats -= int128(amountSats);
+        o.stakedTok -= portionOfStake;
+
+        // Delete the order if its been filled.
+        if (o.amountSats == 0) {
+          delete orderbook[orderID];
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:361]: 👍

**Review by dcposch — APPROVED** [review]
Looks great to me!

**Pull request merged**